Issue #14857: Github generate site fails to generate links with anchors

smita1078 opened the pull request:
Issue #14857

smita1078 commented:
Github, generate site

smita1078 commented:
Github, generate site

smita1078 commented:
Github, generate site

smita1078 commented:
Github, generate site

smita1078 commented:
Github, generate site

smita1078 commented:
Github, generate site

smita1078 commented:
Github, generate site

smita1078 commented:
Github, generate site

smita1078 commented:
Github, generate website

smita1078 commented:
Github, generate website

smita1078 commented:
Github, generate website

romani commented:
I hope you know that you can build website on local 
https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations#how-to-generate-website-only

smita1078 commented:
> I hope you know that you can build website on local https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations#how-to-generate-website-only

Yes, but the issue is with the link that Github bot generates as per my undestanding, Is it possible any way that we can generate this via local?

smita1078 commented:
Also,

```
 while read -r CURRENT_LINE
  do
    # When the line contains 'id='.
    if [[ $CURRENT_LINE =~ id\= ]]
    then
      # Extract the id value from the line.
      SUBSECTION_ID=$(echo "$CURRENT_LINE" | grep -Eo 'id="[^"]+"' | sed 's/id="\([^"]*\)"/\1/')
      echo "SUBSECTION_ID=$SUBSECTION_ID"
      break
    fi
```
as mentioned in the issue, the code here looks for id , I saw that the one for which it generated the report on inspecting had this structure:
```
<section id="Properties"><a id="Properties"></a>
<h2>Properties<div class="anchor"><a href="https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/f0674d6_20250312232802/checks/imports/importorder.html#Properties"><img src="../../images/anchor.png"></a></div></h2>
```

where as
Pattern[] has,

```
<section><a id="Pattern.5B.5D"></a>
<h1> Pattern[] <div class="anchor"><a href="https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/f0674d6_20250312232802/property_types.html#_Pattern[]_"><img src="./images/anchor.png"></a></div></h1>
```

So I modified this to have the similar structure:

```
<section id="Pattern.5B.5D"><a id="Pattern.5B.5D"></a>
<h1>Pattern[]<div class="anchor"><a href="https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a4044d8_20250313033616/property_types.html#Pattern[]"><img src="./images/anchor.png"></a></div></h1>
```
Still on generating website it is not able to generate correct link, Is there anything I am missing, I also introduced few changes so that git diff generates the link , even tried introducing id to section where special characters are not there.. also didn't generate link for this..

```
<section id="TrailingArrayCommaOption"><a id="TrailingArrayCommaOption"></a>
<h1> TrailingArrayCommaOption <div class="anchor"><a href="https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a4044d8_20250313033616/property_types.html#_TrailingArrayCommaOption_"><img src="./images/anchor.png"></a></div></h1>
        
<p> forcing change </p>      
```

romani commented:
Please apply this command on your local https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=27029&view=logs&j=6a833393-c551-5000-0916-5000fd3b86e7&t=bdbbd129-77f0-5d73-ce7d-db58d34c8358&l=46

smita1078 commented:
Github, generate site

smita1078 commented:
![image](https://github.com/user-attachments/assets/120ee7d9-e0c6-48c4-ad93-274072d14e72)

In my local, I am getting the correct link being generated in .ci-temp/messages..
![image](https://github.com/user-attachments/assets/4b255e7f-6c83-48ec-925c-f9deb505f816)

I am unable to figure out why this is not working here via github action bot generate site command.

smita1078 commented:
Github, generate website

smita1078 commented:
Github, generate site

smita1078 commented:
> I hope you know that you can build website on local 

Correct me if I'm wrong, but since the PR isn’t merged, the workflow may be using the repository’s main code — which lacks the modifications— so the link diff. is not visible here. I tried on local and it is generating correct link for me, with proper anchors.

smita1078 commented:
> In my local, I am getting the correct link being generated in .ci-temp/messages.. ![image](https://private-user-images.githubusercontent.com/98531258/422245620-4b255e7f-6c83-48ec-925c-f9deb505f816.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE5ODI0NDIsIm5iZiI6MTc0MTk4MjE0MiwicGF0aCI6Ii85ODUzMTI1OC80MjIyNDU2MjAtNGIyNTVlN2YtNmM4My00OGVjLTkyNWMtZjlkZWI1MDVmODE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE0VDE5NTU0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkM2NhNTg2ZGIzZWQ0YjU2ZmVlMjYyMmY0MmRjODBhZGMzMjMxYWNhMjBmMmZkNGNlNzZjNDg5YjVlYWQwY2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HFNITdb5olWk9GNlmP3xmhuRtibFBK9a7O_E7DsaJOg)
>

romani commented:
ok, if on local works well, but not in amazon, yes, less use github action.

smita1078 commented:
> ok, if on local works well, but not in amazon, yes, less use github action.

Since it's working locally, what should be the next steps for the PR? Do I need do any changes or further debugging? Please Let me know how to proceed further..

romani commented:
as you need help, please move forward to other PRs.

@Zopsss , can you help us at this PR ?

smita1078 commented:
> as you need help, please move forward to other PRs.

Sure, As of now, all my other PRs are merged. This one is in progress, and I'm planning to take up another issue in the meantime

Zopsss commented:
@smita1078 please check this comment: https://github.com/checkstyle/checkstyle/issues/14857#issuecomment-2744310917 and my PR: https://github.com/checkstyle/checkstyle/pull/16624, they might help you

smita1078 commented:
> @smita1078 please check this comment: [#14857 (comment)](https://github.com/checkstyle/checkstyle/issues/14857#issuecomment-2744310917) and my PR: #16624, they might help you

Done!

So actually as mentioned in the issue, when we comment ` Github, generate Site` then it should generate the exact link where changes are made, like `property_types.html#Pattern.5B.5D` this, but instead it is generating `property_types.html# ` this way...

as mentioned in comment , [https://github.com/checkstyle/checkstyle/pull/14740#issuecomment-2089316239](url) 
So I made the changes accordingly in generate-extra-site-link.sh and when I tested in local it did generate the correct link..
In my local, I am getting the correct link being generated in .ci-temp/messages..

![image](https://github.com/user-attachments/assets/4443c7e2-841a-4f92-902a-77af9232af38)

But, when I try to generate links here via comment it does not. As per my understanding the workflow may be using the repository’s main code — which lacks the modifications— so the link diff. is not visible here. But I am not sure on how to proceed further with this, if you could guide me on this..

romani commented:
@Zopsss , please finish review

Zopsss commented:
github, generate site

Zopsss commented:
> As per my understanding the workflow may be using the repository’s main code — which lacks the modifications— so the link diff. is not visible here.

yup that might be the case, once this PR is merged, we will try to generate the site again and see if the anchors links are generating properly or not. You said links were generating locally so I think it should work for `Github, generate site` as well.

Zopsss commented on .ci/generate-extra-site-links.sh:
```diff
@@ -43,37 +43,73 @@ PR_DIFF=$(curl --fail-with-body -s \
+  echo "$DIFF_CONTEXT"
+
+  # Attempt 1: Check for an id on either a <subsection> or <section>
+  SUBSECTION_ID=$(
+    echo "$DIFF_CONTEXT" | grep -Ei '<(subsection|section)' | \
+    grep -Eo 'id="[^"]+"' | head -1 | \
+    sed 's/id="\([^"]*\)"/\1/' | \
+    sed 's/^[[:space:]]*//;s/[[:space:]]*$//'
+  )
```
here you're trying to get ID of `<subsection>` or `<section>`, but our xdoc's sections and subsections doesn't have them. As I explained in my comment: https://github.com/checkstyle/checkstyle/issues/14857#issuecomment-2744310917, the anchor links depends on `<a>` tag present before `<h1>` tag. 

To make your approach work we will need to add ID to all sections and subsections

smita1078 replied:
Yes, for just testing it I added Id to one of the pattern[] section

Zopsss commented on .ci/generate-extra-site-links.sh:
```diff
@@ -43,37 +43,73 @@ PR_DIFF=$(curl --fail-with-body -s \
+    echo "Found id in diff context: $SUBSECTION_ID"
+  fi
+
+  # Fallback: If no id is found, check for the section's name attribute.
+  if [[ -z "$SUBSECTION_ID" ]]; then
+    echo "No id found; checking for section name attribute..."
+    SECTION_NAME=$(
+      echo "$DIFF_CONTEXT" | grep -Ei 'name="[^"]+"' | head -1 | \
+      sed 's/name="\([^"]*\)"/\1/'
+    )
+    SECTION_NAME=$(echo "$SECTION_NAME" | sed 's/^[[:space:]]*//;s/[[:space:]]*$//')
+    echo "SECTION_NAME found: '$SECTION_NAME'"
```
here you're extracting section's name if the ID is not present. This will work for most of the sections but it will fail for the cases I mentioned here: https://github.com/checkstyle/checkstyle/issues/14857#issuecomment-2744310917.

It will fail for cases like byte[], int[], float[], etc because their section name is `Byte[]`, `Int[]`, `Float[]`, etc but their anchor IDs are `byte.5B.5D`, `int.5B.5D`, `float.5B.5D`, etc. 

By `fail`, I mean that it will generate invalid links

Zopsss replied:
Locally you're able to generate anchor link for `Pattern[]` because you added `id` to its `<section>` tag:

https://github.com/checkstyle/checkstyle/pull/16530/files#diff-0b3190eab247f9cc705de22f87e47278c310200ad7d967e59b1cde55a15552f0L677-R677

but as I said in my first review comment, our section and subsection tags doesn't have id, so to generate proper anchor links for `Byte[]`, `Int[]`, `Float[]`, etc sections, we will need to add ID to all of them which I think is not a good approach.

I mean adding IDs to them helps us to generate valid anchor links for them but I think this is a very hacky solution. If you have any better idea please let us know, otherwise we don't have any option and we will have to continue with this approach for now, we can think better solution in the future.

smita1078 replied:
Sure, I will some different approach

Zopsss reviewed (commented):
Items to discuss:

Zopsss commented:
@smita1078 if we decide to continue with `adding ID to section & subsection tags` approach, then please add IDs to all cases which I mentioned here: https://github.com/checkstyle/checkstyle/pull/16530#discussion_r2009063722

smita1078 commented:
> @smita1078 if we decide to continue with `adding ID to section & subsection tags` approach, then please add IDs to all cases which I mentioned here: [#16530 (comment)](https://github.com/checkstyle/checkstyle/pull/16530#discussion_r2009063722)

Yes, Trying some different approach if that's not possible then will add Id to other section & subsection tags as well

smita1078 commented:
> > @smita1078 if we decide to continue with `adding ID to section & subsection tags` approach, then please add IDs to all cases 

New Approach: 
Primary Logic:
The script first reads the diff context from the modified xdoc file and attempts to extract the section tag name attribute using a regular expression. If the name contains special characters, it encodes the string using a jq‑based URI encoder and replaces percent signs with periods to produce a valid anchor to generated URL. (double[] is converted into an anchor such as double.5B.5D)

Fallback Logic: (Existing code)
If no section name is found in the diff context, the script then looks for an anchor id (i.e. an id attribute) within the same context. If an id is present, it uses that value as the anchor. 

Tried this by introducing forced changes in two files... without adding any Id to section or subsection...
Output in local:

![image](https://github.com/user-attachments/assets/cd34b5a9-92ec-44db-83d5-f1b4991b8eff)

in terminal:
```
  xsi:schemaLocation="http://maven.apache.org/XDOC/2.0 https://maven.apache.org/xsd/xdoc-2.0.xsd">
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
<document xmlns="http://maven.apache.org/XDOC/2.0"

<?xml version="1.0" encoding="UTF-8"?>
SECTION_NAME found: 'double[]'
Section name has special characters; encoding...
Derived SUBSECTION_ID: double.5B.5D
CURRENT_XDOC_NAME=property_types
Added link: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0c4c95f_20250305200208/property_types.html#double.5B.5D
Processing file: src/site/xdoc/report_issue.xml
EARLIEST_CHANGE_LINE_NUMBER=13
DIFF_CONTEXT:
      <p>forcing change</p>
    <section name="Content">
  <body>

  </head>
    <title>How to report an issue ?</title>
  <head>

  xsi:schemaLocation="http://maven.apache.org/XDOC/2.0 https://maven.apache.org/xsd/xdoc-2.0.xsd">
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
<document xmlns="http://maven.apache.org/XDOC/2.0"

<?xml version="1.0" encoding="UTF-8"?>
SECTION_NAME found: 'Content'
Section name is alphanumeric; using raw value.
Derived SUBSECTION_ID: Content
CURRENT_XDOC_NAME=report_issue
Added link: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/0c4c95f_20250305200208/report_issue.html#Content

```
Working for both simple text as well as with special characters " [ ] "

Zopsss commented on .ci/generate-extra-site-links.sh:
```diff
@@ -43,37 +43,67 @@ PR_DIFF=$(curl --fail-with-body -s \
+  echo "$DIFF_CONTEXT"
+
+  # Primary: Extract the section's name attribute using grep -Po.
+  SECTION_NAME=$(echo "$DIFF_CONTEXT" | grep -Po '(?<=name=")[^"]+(?=")' | head -1)
```
Here you're extracting `name` of any tag that we encounter, instead of this, can you modify the regex and extract name of only section & subsection? 

It'll make sure that we don't generate any unnecessary links

smita1078 replied:
Yes Sure

Zopsss commented on .ci/generate-extra-site-links.sh:
```diff
@@ -43,37 +43,67 @@ PR_DIFF=$(curl --fail-with-body -s \
+  if [[ -n "$SECTION_NAME" ]]; then
+    if echo "$SECTION_NAME" | grep -q '[^A-Za-z0-9_-]'; then
+      echo "Section name has special characters; encoding..."
+      # Remove newline characters before encoding to avoid unwanted .0A
+      SUBSECTION_ID=$(echo "$SECTION_NAME" | tr -d '\n\r' | jq -sRr @uri | sed 's/%/./g')
```
I'm little confused by this code, can you give an example for it?

smita1078 replied:
`jq -sRr @uri ` ,this portion basically encode the section name if special characters are found..
For example,` double[] → double%5B%5D`
`sed 's/%/./g'` , this  replaces all `% `characters with `. `
`double%5B%5D becomes double.5B.5D`

`tr -d '\n\r`'  , this will remove any new line characters the might be included in section name.. because earlier i was getting the encoding as `double.5B.5D.0A` instead of `double.5B.5D`, I found out `.0A` is hex for new line therefore I added it to be removed and now we are able to get correct encoding.

Zopsss commented on .ci/generate-extra-site-links.sh:
```diff
@@ -43,37 +43,67 @@ PR_DIFF=$(curl --fail-with-body -s \
+    fi
+    echo "Derived SUBSECTION_ID: $SUBSECTION_ID"
+  else
+    # Fallback: If no section name is found, try for an anchor id.
```
@romani please correct me if I'm wrong

romani replied:
Let's avoid, if that is not required

smita1078 replied:
Okay👍

Zopsss reviewed (changes requested):
Great Approach! We're almost there to fix the bug 🪲

Zopsss commented:
> Working for both simple text as well as with special characters " [ ] "

Can you try to find some more examples for special characters other than `[]` ? Just to ensure that our logic is right and we don't generate invalid links

smita1078 commented:
> > Working for both simple text as well as with special characters " [ ] "
> 
> Can you try to find some more examples for special characters other than `[]` ? Just to ensure that our logic is right and we don't generate invalid links

Sure, I will introduce forced change in few other files and will check if its working as expected

Zopsss commented on .ci/generate-extra-site-links.sh:
```diff
@@ -43,37 +43,69 @@ PR_DIFF=$(curl --fail-with-body -s \
+      SUBSECTION_ID=${SECTION_NAME// /_}
+    fi
+
+    # If the anchor starts with a digit, prefix it with an "a".
+    if [[ "$SUBSECTION_ID" =~ ^[0-9] ]]; then
+      SUBSECTION_ID="a$SUBSECTION_ID"
```
Why are we adding prefix "a"?

Zopsss replied:
Good! 

We've links in google style coverage tables which starts with number and they're also prefixed with "a":

https://checkstyle.org/google_style.html#a2.3.1

Though these links are not working currently. These links became broken after we upgraded to latest maven-site-plugin. 

This was not related to your PR I just felt like sharing an example where anchor links starts with numbers.

smita1078 replied:
> We've links in google style coverage tables which starts with number and they're also prefixed with "a":
> Though these links are not working currently. These links became broken after we upgraded to latest maven-site-plugin.

Yes!, I too found these links..

Zopsss commented on .ci/generate-extra-site-links.sh:
```diff
@@ -43,37 +43,69 @@ PR_DIFF=$(curl --fail-with-body -s \
+  SECTION_NAME=$(echo "$SECTION_NAME" | tr -d '\n\r' | \
+    sed 's/^[[:space:]]*//;s/[[:space:]]*$//')
+
+  # Decode all HTML entities using Perl.
+  SECTION_NAME=$(echo "$SECTION_NAME" | \
+    perl -MHTML::Entities -pe '$_ = decode_entities($_)')
```
Can you explain this part?

Zopsss replied:
Asked ChatGPT about this code, it says it used to decodes HTML entities:

```
&amp; → &
```

But I don't think we need this, section names doesn't have such encodings

smita1078 replied:
Actually, I came across this case:
Any HTML entities (like &#39;) are decoded to their literal characters using Perl’s HTML::Entities.
Example: `<section name="Huh? I can&#39;t figure it out!">`  therefore added this..

smita1078 replied:
`"Huh? I can&#39;t figure it out!" → Huh.3F_I_can.26.2339.3Bt_figure_it_out.21` when I tried to generate link via previous code it generated this anchor which is wrong, it was because of this` &#39;` therfore I modified it to handle such html entities if they are present in name attribute.
`"Huh? I can't figure it out!" → Huh.3F_I_can.27t_figure_it_out.21`  it generated correct link

smita1078 replied:
This is in `writingchecks.xml ` , also in a few other files

Zopsss replied:
Are you referring to: https://checkstyle.org/writingchecks.html#Huh?_I_can't_figure_it_out! ?

This links is not working. Are you sure that your approach will fix it?

Zopsss replied:
If yes then we can use this logic in anchor.js too

smita1078 replied:
> Are you referring to: https://checkstyle.org/writingchecks.html#Huh?_I_can't_figure_it_out! ?
Yes

smita1078 replied:
> This links is not working. Are you sure that your approach will generate valid links for `github, generate site` ?

Let me check, will update accordingly..

Zopsss replied:
To check for the valid anchor link you can check the `id` of preceded `<a>` of that `h1` tag.

smita1078 replied:
> This links is not working. Are you sure that your approach will generate valid links for `github, generate site` ?

on generating site in local,

![image](https://github.com/user-attachments/assets/4d635160-bdf6-4059-8674-9bf9341a9f33)

Correct link is being generated, and also new tab is opened successfully!

![image](https://github.com/user-attachments/assets/628f6cbe-4271-4efa-a252-3140e2e1fbb9)

Earlier the link was wrong, as you can see here..

![image](https://github.com/user-attachments/assets/d2dce8d2-b9a4-4f88-8af1-b0d09ed51000)

But after I added your changes in my PR, the link is working.. I dont think there is a need to modify the anchor.js logic.. plz correct me if I am wrong...

smita1078 replied:
> To check for the valid anchor link you can check for the `id` of preceded `<a>` of the `Huh? I can't seem to figure it out`'s `h1` tag.

Yes yes

smita1078 replied:
also in local it genrated the correct link..  (line no. 10)
![image](https://github.com/user-attachments/assets/adc75750-bff6-42fc-97d4-245451f12112)

Zopsss replied:
Gotcha! If that link is working fine then we don't need to change anything :D

smita1078 replied:
Yes :)

smita1078 commented:
> Can you try to find some more examples for special characters other than `[]` ? Just to ensure that our logic is right and we don't generate invalid links

Modifications:
1. The script uses a regex to capture only the name attribute from <section> or <subsection> tags in the diff context.
2. Any HTML entities `(like &#39;)` are decoded to their literal characters using Perl’s HTML::Entities.
Example:  `<section name="Huh? I can&#39;t figure it out!">`

```
<section><a id="Huh.3F_I_can.27t_figure_it_out.21"></a>
<h1>Huh? I can't figure it out!<div class="anchor"><a href="http://localhost:63342/checkstyle/target/site/writingchecks.html#Huh.3F_I_can.27t_figure_it_out.21"><img src="./images/anchor.png"></a></div></h1>
```

3. If special characters are present:
The entire string is URI-encoded (with jq -sRr @uri), then % is replaced with ., and .20 (encoded spaces) is replaced with _.
Otherwise:
Only spaces are replaced with underscores using bash parameter expansion. 
```
<section><a id="Writing_FileSetChecks"></a>
<h1>Writing FileSetChecks<div class="anchor"><a href="http://localhost:63342/checkstyle/target/site/writingchecks.html#Writing_FileSetChecks"><img src="./images/anchor.png"></a></div></h1>

```
4. If the resulting anchor starts with a digit, the script prefixes it with an a (e.g., 4.2E_... → a4.2E_...).
```
<section><a id="a1._Using_Configuration_File_.28-c.2C_--configurationFile.29"></a>
<h4 id="Using_Configuration_File"><span>
1. Using Configuration File (<span class="no-transform">-c, --configurationFile</span>)
</span></h4>
```


Different cases:
name="`double[]`" ---> `#double.5B.5D`
name="`Release 10.12.4`" ---> `#Release_10.12.4`
name="`Content`" ---> `#Content`
name="`Understanding the visitor pattern`" --->` #Understanding_the_visitor_pattern`
name= "`Huh? I can't figure it out!?`" ---> `#Huh.3F_I_can.27t_figure_it_out.21`
name="`Declare check's external resource locations`" ---> `#Declare_check.27s_external_resource_locations`


Output in local:
![image](https://github.com/user-attachments/assets/27d4a567-d0f5-4453-816e-cf2ba2385433)

Zopsss reviewed (approved):
Thanks a lot for your huge effort!!! @smita1078 :D

Everything seems to be working fine now! I think proper anchor links will start to generate after this PR is merged. 

Just revert the forced changes you made in documentation.

Zopsss commented:
@romani please review the PR once again, we changed the approach after your approval.

romani commented:
GitHub, generate website

romani reviewed (approved):
thanks alot!!

it is hard to validate this update fully, as it is internal, we can improve in followup PRs if any defect is found.

smita1078 commented:
> thanks alot!!
> 
> it is hard to validate this update fully, as it is internal, we can improve in followup PRs if any defect is found.

Sure, Thankyou!
